File: workbook/ch08/data/avl/avl.js

class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
        this.height = 1;
    }
}

class AVLTree {
    constructor() {
        this.root = null;
    }

    getHeight(node) {
        return node ? node.height : 0;
    }

    updateHeight(node) {
        node.height = Math.max(this.getHeight(node.left), this.getHeight(node.right)) + 1;
    }

    getBalanceFactor(node) {
        return node ? this.getHeight(node.left) - this.getHeight(node.right) : 0;
    }

    rotateRight(y) {
        const x = y.left;
        const T2 = x.right;

        x.right = y;
        y.left = T2;

        this.updateHeight(y);
        this.updateHeight(x);

        return x;
    }

    rotateLeft(x) {
        const y = x.right;
        const T2 = y.left;

        y.left = x;
        x.right = T2;

        this.updateHeight(x);
        this.updateHeight(y);

        return y;
    }

    insert(node, value) {
        if (!node) return new Node(value);

        if (value < node.value) {
            node.left = this.insert(node.left, value);
        } else if (value > node.value) {
            node.right = this.insert(node.right, value);
        } else {
            return node; // dups not allowed
        }

        this.updateHeight(node);
        return this.rebalance(node);
    }

    delete(node, value) {
        if (!node) return null;

        if (value < node.value) {
            node.left = this.delete(node.left, value);
        } else if (value > node.value) {
            node.right = this.delete(node.right, value);
        } else {
            if (!node.left || !node.right) {
                node = node.left || node.right; // if only one child (or none)
            } else {
                const temp = this.findMin(node.right);
                node.value = temp.value;
                node.right = this.delete(node.right, temp.value);
            }
        }

        if (!node) return null; // if tree empty

        this.updateHeight(node);
        return this.rebalance(node);
    }

    rebalance(node) {
        const balance = this.getBalanceFactor(node);

        // Left Heavy (LL or LR)
        if (balance > 1) {
            if (this.getBalanceFactor(node.left) < 0) {
                node.left = this.rotateLeft(node.left); // LR case
            }
            return this.rotateRight(node); // LL case
        }

        // Right Heavy (RR or RL)
        if (balance < -1) {
            if (this.getBalanceFactor(node.right) > 0) {
                node.right = this.rotateRight(node.right); // RL case
            }
            return this.rotateLeft(node); // RR case
        }

        return node;
    }

    findMin(node) {
        while (node.left) {
            node = node.left;
        }
        return node;
    }

    insertValue(value) {
        this.root = this.insert(this.root, value);
    }

    deleteValue(value) {
        this.root = this.delete(this.root, value);
    }

    visualize() {
        const container = document.getElementById('tree-container');
        container.innerHTML = '';

        if (!this.root) return;

        // maximum depth of the tree
        const maxDepth = this.getHeight(this.root);

        // horizontal spacing based on depth
        const baseOffsetX = (window.innerWidth / Math.pow(2, maxDepth)) * 0.8;
        const baseOffsetY = 80;

        this._visualizeNode(this.root, container, window.innerWidth / 2, 50, baseOffsetX, baseOffsetY);
    }

    _visualizeNode(node, parentElement, x, y, offsetX, offsetY) {
        if (!node) return;

        const nodeElement = document.createElement('div');
        nodeElement.className = 'node';
        nodeElement.textContent = node.value;
        nodeElement.style.left = `${x}px`;
        nodeElement.style.top = `${y}px`;
        parentElement.appendChild(nodeElement);

        if (node.left) {
            const leftX = x - offsetX;
            const leftY = y + offsetY;
            this._drawLine(x + 20, y + 40, leftX + 20, leftY, parentElement);
            this._visualizeNode(node.left, parentElement, leftX, leftY, offsetX / 2, offsetY);
        }

        if (node.right) {
            const rightX = x + offsetX;
            const rightY = y + offsetY;
            this._drawLine(x + 20, y + 40, rightX + 20, rightY, parentElement);
            this._visualizeNode(node.right, parentElement, rightX, rightY, offsetX / 2, offsetY);
        }
    }

    _drawLine(x1, y1, x2, y2, parentElement) {
        const line = document.createElement('div');
        line.className = 'line';
        const length = Math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2);
        const angle = Math.atan2(y2 - y1, x2 - x1) * (180 / Math.PI);
        line.style.width = `${length}px`;
        line.style.transform = `rotate(${angle}deg)`;
        line.style.left = `${x1}px`;
        line.style.top = `${y1}px`;
        parentElement.appendChild(line);
    }
}

const avlTree = new AVLTree();

function addNode() {
    const value = parseInt(document.getElementById('input').value);
    if (!isNaN(value)) {
        avlTree.root = avlTree.insert(avlTree.root, value);
        avlTree.visualize();
    }
}

function deleteNode() {
    const value = parseInt(document.getElementById('input').value);
    if (!isNaN(value)) {
        avlTree.root = avlTree.delete(avlTree.root, value);
        avlTree.visualize();
    }
}

// Zoom and Pan
let scale = 1;
let offsetX = 0;
let offsetY = 0;
let isDragging = false;
let startX, startY;

const treeContainer = document.getElementById('tree-container');

treeContainer.addEventListener('wheel', (e) => {
    e.preventDefault();
    scale += e.deltaY * -0.01;
    scale = Math.min(Math.max(0.5, scale), 2); // scale between 0.5x and 2x
    treeContainer.style.transform = `translate(${offsetX}px, ${offsetY}px) scale(${scale})`;
});

treeContainer.addEventListener('mousedown', (e) => {
    isDragging = true;
    startX = e.clientX - offsetX;
    startY = e.clientY - offsetY;
});

treeContainer.addEventListener('mousemove', (e) => {
    if (isDragging) {
        offsetX = e.clientX - startX;
        offsetY = e.clientY - startY;
        treeContainer.style.transform = `translate(${offsetX}px, ${offsetY}px) scale(${scale})`;
    }
});

treeContainer.addEventListener('mouseup', () => {
    isDragging = false;
});

treeContainer.addEventListener('mouseleave', () => {
    isDragging = false;
});

function zoomIn() {
    scale += 0.1;
    scale = Math.min(scale, 2); // max zoom 2x
    treeContainer.style.transform = `translate(${offsetX}px, ${offsetY}px) scale(${scale})`;
}

function zoomOut() {
    scale -= 0.1;
    scale = Math.max(scale, 0.5); // min zoom 0.5x
    treeContainer.style.transform = `translate(${offsetX}px, ${offsetY}px) scale(${scale})`;
}

File: workbook/ch07/data/common/avl.c
#include <stdio.h>
#include <stdlib.h>

typedef struct AVLNode {
    int key;
    struct AVLNode *left, *right;
    int height;
} AVLNode;

int getHeight(AVLNode *node) {
    return node ? node->height : 0;
}

int getBalance(AVLNode *node) {
    return node ? getHeight(node->left) - getHeight(node->right) : 0;
}

AVLNode *createNode(int key) {
    AVLNode *node = (AVLNode *)malloc(sizeof(AVLNode));
    node->key = key;
    node->left = node->right = NULL;
    node->height = 1;
    return node;
}

AVLNode *rotateRight(AVLNode *z) {
    AVLNode *y = z->left;
    AVLNode *T3 = y->right;
    y->right = z;
    z->left = T3;
    z->height = 1 + (getHeight(z->left) > getHeight(z->right) ? getHeight(z->left) : getHeight(z->right));
    y->height = 1 + (getHeight(y->left) > getHeight(y->right) ? getHeight(y->left) : getHeight(y->right));
    return y;
}

AVLNode *rotateLeft(AVLNode *z) {
    AVLNode *y = z->right;
    AVLNode *T2 = y->left;
    y->left = z;
    z->right = T2;
    z->height = 1 + (getHeight(z->left) > getHeight(z->right) ? getHeight(z->left) : getHeight(z->right));
    y->height = 1 + (getHeight(y->left) > getHeight(y->right) ? getHeight(y->left) : getHeight(y->right));
    return y;
}

AVLNode *insert(AVLNode *node, int key) {
    if (!node) return createNode(key);

    if (key < node->key)
        node->left = insert(node->left, key);
    else
        node->right = insert(node->right, key);

    node->height = 1 + (getHeight(node->left) > getHeight(node->right) ? getHeight(node->left) : getHeight(node->right));
    int balance = getBalance(node);

    if (balance > 1 && key < node->left->key)
        return rotateRight(node);
    if (balance < -1 && key > node->right->key)
        return rotateLeft(node);
    if (balance > 1 && key > node->left->key) {
        node->left = rotateLeft(node->left);
        return rotateRight(node);
    }
    if (balance < -1 && key < node->right->key) {
        node->right = rotateRight(node->right);
        return rotateLeft(node);
    }

    return node;
}

void inorder(AVLNode *root) {
    if (root) {
        inorder(root->left);
        printf("%d ", root->key);
        inorder(root->right);
    }
}

int main() {
    AVLNode *root = NULL;
    int keys[] = {10, 20, 30, 40, 50, 25};
    
    for (int i = 0; i < 6; i++)
        root = insert(root, keys[i]);

    printf("Inorder traversal after insertions:\n");
    inorder(root);
    printf("\n");

    return 0;
}